frontend/src/utils/lazyRetry.ts

```diff
@@ -0,0 +1,36 @@
+import { ComponentType } from 'react'
+
+/**
+ * Retry React.lazy calls on error, or refresh the page if it was a chunk load
+ * error to hopefully retrieve the correct version of the chunk to due
+ * no-cache header.
+ *
+ * Retrieved from https://www.codemzy.com/blog/fix-chunkloaderror-react.
+ *
+ * @note this only works for route based code splitting https://reactjs.org/docs/code-splitting.html#route-based-code-splitting
+ * Will cause infinite loop if used for normal code splitting since multiple lazy calls may be triggered in a single component.
+ */
+export const lazyRetry = <T extends ComponentType>(
+  componentImport: () => Promise<{ default: T }>,
+) => {
+  return new Promise<{ default: T }>((resolve, reject) => {
+    // check if the window has already been refreshed
+    const hasRefreshed = JSON.parse(
+      window.sessionStorage.getItem('retry-lazy-refreshed') || 'false',
+    )
+    // try to import the component
+    componentImport()
+      .then((component) => {
+        window.sessionStorage.setItem('retry-lazy-refreshed', 'false') // success so reset the refresh
+        resolve(component)
+      })
+      .catch((error) => {
+        if (!hasRefreshed) {
+          // not been refreshed yet
+          window.sessionStorage.setItem('retry-lazy-refreshed', 'true') // we are now going to refresh
```

Comment:
Why do we need to store this in `sessionStorage`? `sessionStorage` survives a page refresh, but the `hasRefreshed` flag is meant to apply only for the current instance of the app, correct?

Reply:
makes more sense to store it here than in localStorage, no? but you are right.. only happens for a single deploy...

wonder if......... we should just refresh if the version of the app has changed......

Reply:
Well, I reckon it shouldn't be in sessionstorage or localstorage; it should just be in app memory. A singleton flag would do.

Reply:
if its in memory won't it be cleared after the first refresh? then it'll just lead to an infinite refresh loop if the chunk just 404s forever

Reply:
Oh you meant to do a single refresh overall? And if still broken, user would refresh manually thereafter?

I thought we would try to refresh till it works (under the assumption is eventually will work). In the cases of deployments (with vN and vN+1 existing in parallel), we could have multiple consecutive failures, but they would eventually resolve.

But you're right! If we have a broken deployment, and all clients go into refresh frenzy, they might DDOS us 😰. For continuous refresh, we should have some backoff mechanism.

Reply:
single refresh should account for the bulk of the errors... but this would be a single refresh for the lifetime of the tab... which doesn't mean much. Maybe we should update once per version, which would already exist with `process.env.REACT_APP_VERSION`!

Reply:
done in [bbb0eb1](https://github.com/opengovsg/FormSG/pull/5270/commits/bbb0eb1d298d543ba4ca95bb4ae6a5ea54b7874f)

Reply:
Aye, that would be nice 👍